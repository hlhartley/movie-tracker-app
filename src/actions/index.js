export const addMovies = (movies) => ({
  type: 'ADD_MOVIES',
  movies
});

export const loginUser = (id, name) => ({
  type: 'LOGIN_USER',
  id,
  name
});

export const updateError = (message) => ({
  type: 'UPDATE_ERROR',
  message
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});

export const showPopup = (bool) => ({
  type: 'SHOW_POPUP',
  showPopup: bool
});

export const toggleFavorite = (id) => ({
  type: 'TOGGLE_FAVORITE',
  id
});
