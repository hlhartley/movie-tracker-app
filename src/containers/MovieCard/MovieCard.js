import React from 'react';
// import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

const MovieCard = ({ id, original_title, overview, release_date, vote_average, poster_path }) => {
    return (
        <div className='movie-card'>
            <i class={'fas fa-crown favorite-false'}></i>
            <Link to={`movies/${id}`}>
            <img src={`https://image.tmdb.org/t/p/w500${poster_path}`} className='movie-poster'/>
            </Link>
        </div>
    )
}

export default MovieCard