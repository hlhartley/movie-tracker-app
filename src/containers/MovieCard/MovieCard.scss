.movie-card {
  background-color: $brown;
  border: 3px solid transparent;
  border-radius: 5px;
  width: 20%;
  text-align: center;
  padding: 20px 0 20px 0;
  margin: 25px;
  &:hover {
    border: 3px solid $darksilver;
  }
}

.fa-crown {
  display: block;
  font-size: 20px;
  margin-bottom: 10px;
  position: relative;
}

.favorite-false {
  color: $silver;
  &:hover {
    color: $gold;
  }
}

.favorite-true {
  color: $gold;
}

.movie-poster {
  width: 80%;
}