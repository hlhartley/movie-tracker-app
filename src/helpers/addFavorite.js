export const addFavorite = async (movie_id, user_id, title, poster_path, release_date, vote_average, overview) => {
  const response = await fetch('http://localhost:3000/api/users/favorites/new', {
    method: 'POST',
    body: JSON.stringify({
      movie_id, 
      user_id, 
      title, 
      poster_path, 
      release_date, 
      vote_average, 
      overview
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (response.status >= 300) {
    throw Error(`Error saving to favorites: ${response.statusText}`);
  } else {
    const result = await response.json();
    return result;
  }
}