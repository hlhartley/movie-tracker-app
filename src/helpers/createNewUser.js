export const createNewUser = async (user) => {

  const response = await fetch('http://localhost:3000/api/users/new', {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (response.status >= 300) {
    throw Error(`Email has already been used: ${response.statusText}`);
  } else {
    const result = await response.json();
    return result;
  }

}