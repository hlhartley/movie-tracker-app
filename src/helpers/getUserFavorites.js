export const getUserFavorites = async (user_id) => {
    const response = await fetch(`http://localhost:3000/api/users/${user_id}/favorites`)
    if(response.status >= 300) {
        throw Error(`Error getting favorites: ${response.statusText}`)
    } else {
        const result = response.json()
        return result
    }
}