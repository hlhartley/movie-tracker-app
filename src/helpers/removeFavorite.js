export const removeFavorite = async (user_id, movie_id) => {
  const response = await fetch(`http://localhost:3000/api/users/${user_id}/favorites/${movie_id}`, {
    method: 'DELETE',
    body: JSON.stringify({
      user_id,
      movie_id
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (response.status >= 300) {
    throw Error(`Error removing favorite: ${response.statusText}`)
  } else {
    const result = response.json();
    return result;
  }
}