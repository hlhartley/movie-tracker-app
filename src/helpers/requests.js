import { apikey } from '../apikey';

export const fetchMovies = async () => {
  const url = `https://api.themoviedb.org/3/movie/now_playing?api_key=${apikey}&language=en-US&page=1`;
  const response= await fetch(url);
  if (response.status >= 300) {
    throw new Error(`Error fetching', code: ${response.status}`)
  } else {
    return response.json();
  }
}