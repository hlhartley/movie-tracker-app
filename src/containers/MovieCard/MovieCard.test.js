import React from 'react';
import { shallow } from 'enzyme';
import { MovieCard, mapStateToProps, mapDispatchToProps } from './MovieCard';
import * as Requests from '../../helpers/requests';
import { showPopup, toggleFavorite, addFavorite, removeFavorite } from '../../actions';

describe('MovieContain', () => {
  describe('MovieContain component', () => {
    let mockMovie = {
      vote_count: 3119,
      id: 297802,
      video: false,
      vote_average: 6.9,
      title: "Aquaman",
      popularity: 483.065,
      poster_path: "/5Kg76ldv7VxeX9YlcQXiowHgdX6.jpg",
      original_language: "en",
      original_title: "Aquaman",
      backdrop_path: "/5A2bMlLfJrAfX9bqAibOL2gCruF.jpg",
      adult: false,
      overview: "Once home to the most advanced civilization on Earth, the city of Atlantis is now an underwater kingdom ruled by the power-hungry King Orm. With a vast army at his disposal, Orm plans to conquer the remaining oceanic people -- and then the surface world. Standing in his way is Aquaman, Orm's half-human, half-Atlantean brother and true heir to the throne. With help from royal counselor Vulko, Aquaman must retrieve the legendary Trident of Atlan and embrace his destiny as protector of the deep.",
      release_date: "2018-12-07",
      isFavorite: false
    };
    let mockFavMovie = {
      vote_count: 1020,
      id: 424783,
      video: false,
      vote_average: 6.5,
      title: "Bumblebee",
      popularity: 263.575,
      poster_path: "/fw02ONlDhrYjTSZV8XO6hhU3ds3.jpg",
      original_language: "en",
      original_title: "Bumblebee",
      backdrop_path: "/hMANgfPHR1tRObNp2oPiOi9mMlz.jpg",
      adult: false,
      overview: "On the run in the year 1987, Bumblebee finds refuge in a junkyard in a small Californian beach town. Charlie, on the cusp of turning 18 and trying to find her place in the world, discovers Bumblebee, battle-scarred and broken. When Charlie revives him, she quickly learns this is no ordinary yellow VW bug.",
      release_date: "2018-12-15",
      isFavorite: true
    }

    let mockUser = { id: 2, name: 'Jill'};
    let mockNoUser = null;
    const mockHandleShowPopup = jest.fn();
    const mockToggleFavorite = jest.fn();
    const mockAddFavorite = jest.fn();
    const mockRemoveFavorite = jest.fn();

    beforeEach(() => {
      Requests.addFavoriteToDB = jest.fn(() => Promise.resolve({ message: "Movie was added to favorites" }));
      Requests.removeFavoriteFromDB = jest.fn(() => Promise.resolve({ status: "success"}));
    });

    it('should match the snapshot with all data passed in', () => {
      const wrapper = shallow(<MovieCard {...mockMovie} currentUser={mockUser} handleShowPopup={mockHandleShowPopup} toggleFavorite={mockToggleFavorite} addFavorite={mockAddFavorite} removeFavorite={mockRemoveFavorite} />)
      expect(wrapper).toMatchSnapshot();
    });

    describe('handleClick', () => {
      it('should call removeFavoriteFromDB with the correct params if a user is logged in and the movie is already favorited', async () => {
        const wrapper = shallow(<MovieCard {...mockFavMovie} currentUser={mockUser} handleShowPopup={mockHandleShowPopup} toggleFavorite={mockToggleFavorite} addFavorite={mockAddFavorite} removeFavorite={mockRemoveFavorite} />)
        const expectedUserId = mockUser.id;
        const expectedMovieId = mockFavMovie.id;

        await wrapper.instance().handleClick();
        expect(Requests.removeFavoriteFromDB).toHaveBeenCalledWith(expectedUserId, expectedMovieId)
      });

      it('should call removeFavorite with the correct id if a user is logged in and the movie is already favorited', async () => {
        const wrapper = shallow(<MovieCard {...mockFavMovie} currentUser={mockUser} handleShowPopup={mockHandleShowPopup} toggleFavorite={mockToggleFavorite} addFavorite={mockAddFavorite} removeFavorite={mockRemoveFavorite} />)
        const expectedMovieId = mockFavMovie.id;
        await wrapper.instance().handleClick();
        expect(mockRemoveFavorite).toHaveBeenCalledWith(expectedMovieId);
      });

      it('should call toggleFavorites with the correct id if a user is logged in and the movie is already favorited', async () => {
        const wrapper = shallow(<MovieCard {...mockFavMovie} currentUser={mockUser} handleShowPopup={mockHandleShowPopup} toggleFavorite={mockToggleFavorite} addFavorite={mockAddFavorite} removeFavorite={mockRemoveFavorite} />)
        const expectedMovieId = mockFavMovie.id;
        await wrapper.instance().handleClick();
        expect(mockToggleFavorite).toHaveBeenCalledWith(expectedMovieId)
      });

      it('should call addFavoriteToDB with the correct params if a user is logged in and the movie is not already favorited', async () => {
        const wrapper = shallow(<MovieCard {...mockMovie} currentUser={mockUser} handleShowPopup={mockHandleShowPopup} toggleFavorite={mockToggleFavorite} addFavorite={mockAddFavorite} removeFavorite={mockRemoveFavorite} />);        
        await wrapper.instance().handleClick();
        expect(Requests.addFavoriteToDB).toHaveBeenCalledWith(mockMovie.id, mockUser.id, mockMovie.title, mockMovie.poster_path, mockMovie.release_date, mockMovie.vote_average, mockMovie.overview)
      });

      it('should call addMovie with the corrct id to favorites if a user is logged in and the movie is not already favorited', async () => {
        const wrapper = shallow(<MovieCard {...mockMovie} currentUser={mockUser} handleShowPopup={mockHandleShowPopup} toggleFavorite={mockToggleFavorite} addFavorite={mockAddFavorite} removeFavorite={mockRemoveFavorite} />);        
        const expected = mockMovie.id
        await wrapper.instance().handleClick()
        expect(mockAddFavorite).toHaveBeenCalledWith(expected);
      });

      it('should call toggleFavorite with the correct id if a user is logged in and the movie is not already favorited', async () => {
        const wrapper = shallow(<MovieCard {...mockMovie} currentUser={mockUser} handleShowPopup={mockHandleShowPopup} toggleFavorite={mockToggleFavorite} addFavorite={mockAddFavorite} removeFavorite={mockRemoveFavorite} />);        
        const expected = mockMovie.id;
        await wrapper.instance().handleClick();
        expect(mockToggleFavorite).toHaveBeenCalledWith(expected);
      });

      it('should call handleShowPopup with the correct boolean if there is no user logged in', () => {
        const expected = true;
        const wrapper = shallow(<MovieCard {...mockMovie} currentUser={mockNoUser} handleShowPopup={mockHandleShowPopup} toggleFavorite={mockToggleFavorite} addFavorite={mockAddFavorite} removeFavorite={mockRemoveFavorite} />);
        wrapper.instance().handleClick();
        expect(mockHandleShowPopup).toHaveBeenCalledWith(expected);
      });
    })

  });

  describe('mapStateToProps', () => {
    it('should return an object with currentUser', () => {
      const mockState = {
        currentUser: { id: 4, name: 'Mary' },
        errorStatus: '',
        favoriteMovies: [],
        showPopup: false
      }
      const expected = {
        currentUser: { id: 4, name: 'Mary' }
      }
      const result = mapStateToProps(mockState);
      expect(result).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with a showPopup action whem handleShowPopup is called', () => {
      const mockDispatch = jest.fn();
      const bool = true;
      const actionToDispatch = showPopup(bool);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.handleShowPopup(bool);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with a toggleFacorite action when toggleFavorite is called', () => {
      const mockDispatch = jest.fn();
      const movieId = 3467;
      const actionToDispatch = toggleFavorite(movieId);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.toggleFavorite(movieId);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with a addFavorite action when addFavorite is called', () => {
      const mockDispatch = jest.fn();
      const movieId = 6982;
      const actionToDispatch = addFavorite(movieId);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addFavorite(movieId);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with a removeFavorite action when removeFavorite is called', () => {
      const mockDispatch = jest.fn();
      const movieId = 92893;
      const actionToDispatch = removeFavorite(movieId);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.removeFavorite(movieId);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });

})