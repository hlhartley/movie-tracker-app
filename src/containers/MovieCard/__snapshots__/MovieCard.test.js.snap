// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieContain MovieContain component should match the snapshot with all data passed in 1`] = `
<div
  className="movie-card"
>
  <i
    className="fas fa-crown favorite-false"
    onClick={[Function]}
  />
  <Link
    replace={false}
    to="movies/297802"
  >
    <img
      alttext="Aquaman"
      className="movie-poster"
      src="https://image.tmdb.org/t/p/w500/5Kg76ldv7VxeX9YlcQXiowHgdX6.jpg"
    />
  </Link>
</div>
`;
